spring-rabbit/src/main/java/org/springframework/amqp/rabbit/core/AsyncRabbitTemplate.java

```diff
@@ -0,0 +1,583 @@
+			correlationData = new CorrelationData(correlationId);
+			future.setConfirm(new SettableListenableFuture<Boolean>());
+		}
+		this.pending.put(correlationId, future);
+		this.template.send(exchange, routingKey, message, correlationData);
+		return future;
+	}
+
+	/**
+	 * Convert the object to a message and send it to the provided exchange and
+	 * routing key after invoking the {@link MessagePostProcessor}.
+	 * If the post processor adds a correlationId property, it must be unique.
+	 * @param exchange the exchange
+	 * @param routingKey the routing key.
+	 * @param message the message.
+	 * @param messagePostProcessor the post processor.
+	 * @return the {@link RabbitConverterFuture}.
+	 */
+	public RabbitConverterFuture convertSendAndReceive(String exchange, String routingKey, Object message,
+			MessagePostProcessor messagePostProcessor) throws AmqpException {
+		CorrelationData correlationData = null;
+		if (this.enableConfirms) {
+			correlationData = new CorrelationData(null);
+		}
+		CorrelationMessagePostProcessor correlationPostProcessor = new CorrelationMessagePostProcessor(
+				messagePostProcessor, correlationData);
+		this.template.convertAndSend(exchange, routingKey, message, correlationPostProcessor, correlationData);
+		return correlationPostProcessor.getFuture();
+	}
+
+	@Override
+	public synchronized void start() {
+		if (!this.running) {
+			this.container.start();
+		}
+		this.running = true;
+	}
+
+	@Override
+	public synchronized void stop() {
+		if (this.running) {
+			this.container.stop();
+		}
+		this.running = false;
+	}
+
+	@Override
+	public boolean isRunning() {
+		return this.running;
+	}
+
+	@Override
+	public int getPhase() {
+		return this.phase;
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return this.autoStartup;
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		stop();
+		callback.run();
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public void onMessage(Message message) {
+		MessageProperties messageProperties = message.getMessageProperties();
+		if (messageProperties != null) {
+			byte[] correlationId = messageProperties.getCorrelationId();
+			if (correlationId != null) {
+				if (logger.isDebugEnabled()) {
+					logger.debug("onMessage: " + message);
+				}
+				RabbitFuture<?> future = this.pending.remove(new String(correlationId, this.charset));
+				if (future != null) {
+					if (future instanceof AsyncRabbitTemplate.RabbitConverterFuture) {
+						C converted = (C) this.template.getMessageConverter().fromMessage(message);
+						((RabbitConverterFuture) future).set(converted);
+					}
+					else {
+						((RabbitMessageFuture) future).set(message);
+					}
+				}
+			}
+		}
+	}
+
+	@Override
+	public void returnedMessage(Message message, int replyCode, String replyText, String exchange, String routingKey) {
+		String correlationId = new String(message.getMessageProperties().getCorrelationId(), this.charset);
+		if (correlationId != null) {
```

Comment:
It's actually unfortunate that `MessageProperties` uses `byte[]` since `BasicProperties` uses `String` so we end up converting back - see `DefaultMessagePropertiesConverter` - I thought we had a JIRA to fix that (support a String as an alternative to byte[]) but I haven't found it yet.